lib/logstash/outputs/tcp.rb

```diff
@@ -183,29 +216,40 @@ def connect
         begin
           client_socket.connect
         rescue OpenSSL::SSL::SSLError => ssle
-          @logger.error("SSL Error", :exception => ssle, :backtrace => ssle.backtrace)
+          log_error 'connect ssl failure:', ssle, backtrace: false
           # NOTE(mrichar1): Hack to prevent hammering peer
           sleep(5)
           raise
         end
       end
       client_socket.instance_eval { class << self; include ::LogStash::Util::SocketPeer end }
-      @logger.debug("Opened connection", :client => "#{client_socket.peer}")
+      @logger.debug("opened connection", :client => client_socket.peer)
       return client_socket
-    rescue StandardError => e
-      @logger.error("Failed to connect: #{e.message}", :exception => e.class, :backtrace => e.backtrace)
+    rescue => e
+      log_error 'failed to connect:', e
       sleep @reconnect_interval
       retry
     end
   end # def connect
 
-  private
   def server?
     @mode == "server"
   end # def server?
 
-  public
-  def receive(event)
-    @codec.encode(event)
-  end # def receive
+  def log_warn(msg, e, details = {})
+    details = details.merge message: e.message, exception: e.class
+    if details[:backtrace] || (details[:backtrace].nil? && @logger.debug?)
```

Comment:
`:backtrace` here seems to be doing two things: first it's used as a predicate, then it's used as a key to store an exceptions backtrace.
Also the two `log_warn` and `log_error` behave differently, one takes [String, Exception, Hash], and another [String, Exception, Boolean].
Can we just make it `[String, Exception, Boolean, Hash]` for both?